### ISSUE TITLE
SudoSpawner stopped working after the upgrade

### ISSUE THREAD

**Issue by hani1814** [jupyterhub/jupyterhub#1388]
Hi,

I had jupyterhub / spawner running for some time under a service account. I have upgraded our test platform to 0.8.0b3 and I can no longer run under a service account.

Worth noting that we have modified the file spawner.py to run under dzdo (centrify) as we are not allowed to change the sudoer file. The command is:
cmd = ['dzdo', '-u', self.user.name]

The above has been working for over a year now

The following scenarios were tested:

1-It works if I start jupyterhub as root
2-Fails if I start it under the service account (which always worked)

This is the error message I am getting

```
[I 2017-09-01 13:02:07.571 JupyterHub proxy:458] Starting proxy @ https://*:9996/
13:02:07.739 - info: [ConfigProxy] Proxying https://*:9996 to (no default)
13:02:07.741 - info: [ConfigProxy] Proxy API at http://127.0.0.1:8001/api/routes
[W 2017-09-01 13:02:08.451 JupyterHub proxy:304] Adding missing default route
[I 2017-09-01 13:02:08.452 JupyterHub proxy:370] Adding default route for Hub: / => http://127.0.0.1:9997
13:02:08.453 - info: [ConfigProxy] 200 GET /api/routes
13:02:08.460 - info: [ConfigProxy] Adding route / -> http://127.0.0.1:9997
13:02:08.462 - info: [ConfigProxy] 201 POST /api/routes/
[I 2017-09-01 13:02:08.463 JupyterHub app:1584] JupyterHub is now running at https://:9996/
[I 2017-09-01 13:02:47.502 JupyterHub log:122] 302 GET / â /hub (@10.199.195.16) 1.19ms
[I 2017-09-01 13:02:48.201 JupyterHub log:122] 302 GET /hub â /hub/login (@10.199.195.16) 1.26ms
[I 2017-09-01 13:02:48.501 JupyterHub log:122] 200 GET /hub/login (@10.199.195.16) 36.53ms
[I 2017-09-01 13:03:04.679 JupyterHub base:345] User logged in: d3849648
[W 2017-09-01 13:03:04.713 JupyterHub auth:570] Failed to open PAM session for d3849648: [PAM Error 14] Cannot make/remove an entry for the specified session
[W 2017-09-01 13:03:04.713 JupyterHub auth:571] Disabling PAM sessions from now on.
[E 2017-09-01 13:03:04.718 JupyterHub user:425] Unhandled error starting d3849648's server: 'SudoSpawner' object has no attribute 'db'
[E 2017-09-01 13:03:04.727 JupyterHub web:1590] Uncaught exception POST /hub/login?next= (10.199.195.16)
    HTTPServerRequest(protocol='https', host='hkg3pl0242.hk.XXXX:9996', method='POST', uri='/hub/login?next=', version='HTTP/1.1', remote_ip='10.199.195.16', headers={'Origin': 'https://hkg3pl0242.hk.XXXX:9996', 'X-Forwarded-For': '10.199.195.16', 'Upgrade-Insecure-Requests': '1', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8', 'Accept-Encoding': 'gzip, deflate, br', 'Connection': 'close', 'X-Forwarded-Host': 'hkg3pl0242.hk.XXXX:9996', 'X-Forwarded-Proto': 'https', 'Accept-Language': 'en-GB,en-US;q=0.8,en;q=0.6', 'Dnt': '1', 'Referer': 'https://hkg3pl0242.hk.XXXX:9996/hub/login', 'Content-Length': '36', 'Cache-Control': 'max-age=0', 'User-Agent': 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36', 'Host': 'hkg3pl0242.hk.XXXX:9996', 'X-Forwarded-Port': '9996', 'Content-Type': 'application/x-www-form-urlencoded'})
    Traceback (most recent call last):
      File "/opt/app/anaconda3/lib/python3.5/site-packages/tornado/web.py", line 1511, in _execute
        result = yield result
      File "/opt/app/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/login.py", line 93, in post
        yield self.spawn_single_user(user)
      File "/opt/app/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/base.py", line 465, in spawn_single_user
        yield gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), finish_user_spawn())
      File "/opt/app/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/base.py", line 444, in finish_user_spawn
        yield spawn_future
      File "/opt/app/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 436, in spawn
        raise e
      File "/opt/app/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 377, in spawn
        ip_port = yield gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
      File "/opt/app/anaconda3/lib/python3.5/site-packages/sudospawner/spawner.py", line 89, in start
        self.db.commit()
    AttributeError: 'SudoSpawner' object has no attribute 'db'

[E 2017-09-01 13:03:04.738 JupyterHub log:114] {
      "Origin": "https://hkg3pl0242.hk.XXXX:9996",
      "X-Forwarded-For": "10.199.195.16",
      "Upgrade-Insecure-Requests": "1",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
      "Accept-Encoding": "gzip, deflate, br",
      "Connection": "close",
      "X-Forwarded-Host": "hkg3pl0242.hk.XXXX:9996",
      "X-Forwarded-Proto": "https",
      "Accept-Language": "en-GB,en-US;q=0.8,en;q=0.6",
      "Dnt": "1",
      "Referer": "https://hkg3pl0242.hk.XXXX:9996/hub/login",
      "Content-Length": "36",
      "Cache-Control": "max-age=0",
      "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36",
      "Host": "hkg3pl0242.hk.XXXX:9996",
      "X-Forwarded-Port": "9996",
      "Content-Type": "application/x-www-form-urlencoded"
    }
[E 2017-09-01 13:03:04.738 JupyterHub log:122] 500 POST /hub/login?next= (@10.199.195.16) 112.45ms
```

**Comment by hani1814** [comment]
worth noting that I have tested it by changing /etc/sudoers file

Cmnd_Alias JUPYTER_CMD = /opt/app/anaconda3/bin/sudospawner
hdp50-srv-jup ALL=(%test_u1) NOPASSWD:JUPYTER_CMD

the test works from the command line but I get a very similar error to the above. It appears that there is a missing db which may have not existed in the old version?

**Comment by hani1814** [comment]
I've made progress on this.

To start with, the /opt/app/anaconda3/lib/python3.5/site-packages/sudospawner/spawner.py file was not properly updated to the latest version, so I replaced it but this did not solve the problem.

When I created the config file in ~/.jupyter/jupyter_notebook_config.py it worked

Not sure why and the only changes in there are:
c.NotebookApp.port = 9003
c.NotebookApp.ip = '*'
c.NotebookApp.open_browser = False

Any idea why this setting works? The notebook port appears to be ignored plus notice that the jupyter log now has a reference to SingleUserNotebookApp:
```
[I 2017-09-01 15:37:39.596 SingleUserNotebookApp log:122] 200 GET /user/d3849648/tree? (d3849648@10.199.195.16) 83.48ms
[I 2017-09-01 15:37:41.700 SingleUserNotebookApp log:122] 200 GET /user/d3849648/api/config/tree?_=1504280260674 (d3849648@10.199.195.16) 4.13ms
[I 2017-09-01 15:37:41.704 SingleUserNotebookApp log:122] 200 GET /user/d3849648/api/config/common?_=1504280260675 (d3849648@10.199.195.16) 2.53ms
[I 2017-09-01 15:37:42.413 SingleUserNotebookApp log:122] 200 GET /user/d3849648/api/terminals?_=1504280260676 (d3849648@10.199.195.16) 1.35ms
```

**Comment by minrk** [comment]
This has been fixed by #1384 (0.8.0b4) and https://github.com/jupyterhub/sudospawner/pull/41 . Did you perhaps update to the latest beta while testing and that fixed it?

**Comment by willingc** [comment]
Closing this issue since it appears to be resolved. Please feel free to leave an issue at [jupyter/help](https://github.com/jupyter/help) or leave a comment requesting this issue be reopened. Thanks to all!
